File: prev_exams_and_ans/2018ccc/q4_sunflows/Sunflowers.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Sunflowers {
	private static int[][] grid;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		
		grid = new int[N][N];
		for (int i = 0; i < N; i++) {
			String[] row = br.readLine().split(" ");
			for (int j = 0; j < N; j++) {
				grid[i][j] = Integer.parseInt(row[j]);
			}
		}
		
		transBackFrom(determineDegree());
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < N; j++) {
				System.out.print(grid[i][j] + " ");
			}
			System.out.println();
		}
		
		
	}
	
	public static String determineDegree() {
		int N = grid.length;
		if (grid[0][N-1] < grid[0][N-2] && grid[0][N-1] < grid[1][N-1]) {
			return "positive_90";
		} else if (grid[N-1][N-1] < grid[N-1][N-2] && grid[N-1][N-1] < grid[N-2][N-1]) {
			return "diagonal";
		} else if (grid[N-1][0] < grid[N-2][0] && grid[N-1][0] < grid[N-1][1]) {
			return "negative_90";
		}
		return "zero";
	}
	
	public static void transBackFrom(String direction) {
		int N = grid.length;
		int[][] rotatedGrid = new int[N][N];
		
		if (direction.equals("positive_90")) {
			for(int x = 0; x < N; x++) {
		        for(int y = 0; y < N; y++) {
		            rotatedGrid[N-1-y][x] = grid[x][y];
		        }
		    }
			grid = rotatedGrid;
		} else if (direction.equals("diagonal")) {
			for(int x = 0; x < N; x++) {
		        for(int y = 0; y < N; y++) {
		            rotatedGrid[N-1-x][N-1-y] = grid[x][y];
		        }
		    }
			grid = rotatedGrid;
		} else if (direction.equals("negative_90")) {
		    for(int x = 0; x < N; x++) {
		        for(int y = 0; y < N; y++) {
		            rotatedGrid[y][N-1-x] = grid[x][y];
		        }
		    }
		    grid = rotatedGrid;
		}
		
	}
}

File: prev_exams_and_ans/2018ccc/q4_sunflows/sun_flowers.py
def transback_from(degree, grid):
    N = len(grid)
    rotated_grid = [[0 for x in range(0, N)] for y in range(0, N)]
		
    if degree == "positive_90":
        for x in range(0, N):
            for y in range(0, N):
                rotated_grid[N-1-y][x] = grid[x][y]
    elif degree == "diagonal":
        for x in range(0, N):
            for y in range(0, N):
                rotated_grid[N-1-x][N-1-y] = grid[x][y]
    elif degree == "negative_90":
        for x in range(0, N):
            for y in range(0, N):
                rotated_grid[y][N-1-x] = grid[x][y]
    else: 
        return grid

    return rotated_grid


def determine_degree(grid):
    N = len(grid)
    if (grid[0][N-1] < grid[0][N-2] and grid[0][N-1] < grid[1][N-1]):
        return "positive_90"
    elif (grid[N-1][N-1] < grid[N-1][N-2] and grid[N-1][N-1] < grid[N-2][N-1]):
        return "diagonal"
    elif (grid[N-1][0] < grid[N-2][0] and grid[N-1][0] < grid[N-1][1]):
        return "negative_90"
    
    return "zero";


N = int(input())

grid = []

for i in range(0, N):
    grid.append([int(x) for x in input().split()])

grid = transback_from(determine_degree(grid), grid)

for i in range(0, N):
    for j in range(0, N):
        print(grid[i][j], "", end="")
    print()
